### PR TITLE
Dataloader #1 code review

### PR DESCRIPTION
**Pull request by Uniaut**
이게 몇번째야 으

### CODE REVIEW

File: Dataloader/dataloader.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
+import torch.utils.data as dt
+import json, time
+from PIL import Image, ImageDraw
+import torchvision.transforms
+import torch.random
+import copy
+
+from Utils.Quadangle import Quadangle
```
[the remark this answers is not in the document]
Comment (kozistr) [reply Dataloader/dataloader.py]: quadangle 보다 더 직관적인 클래스 이름 찾아봐도 좋을 듯

[the remark this answers is not in the document]
Comment (Uniaut) [reply Dataloader/dataloader.py]: 사각형이라는 뜻으로 square나 rectangle이 부적절할 것 같아서 임시로 지은 이름인데 데이터셋 확인해보니까 완벽하게 직사각형이어서 rbox같은걸로 고칠 예정

File: Dataloader/dataloader.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    def __init__(self, img_id: str, transform=None, crop=(0, 0, 512, 512)):
+        self.img_id = img_id
+        self.transform = transform
+        self.crop = crop
+
+    def __hash__(self):
+        return hash((self.img_id, self.transform, self.crop))
+
+    def __eq__(self, other):
+        return hash(self) == hash(other)
```
[the remark this answers is not in the document]
Comment (Uniaut) [reply Dataloader/dataloader.py]: 인터넷에서 이렇게 나와있길래... ㅎ;;

File: Dataloader/dataloader.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kozistr) [reply Dataloader/dataloader.py]: ㅇㅎ 그게 더 좋아 보인다 RBox 나 이런 거로 하면 좋을 듯
일단 내가 코드 이해는 안하고 코드만 보긴 했는데 Quandangle 이 정확히 어떤 일을 하는 클래스임?

File: Dataloader/dataloader.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+            if img_id in self.features_map.keys():
+                self.features_map.get(img_id).add_box(q, label)
+            else:
+                # self.features_map[img_id] = Features(img_id)
+                pass
+
+        for img_id, f in self.features_map.items():
+            # print(img_id)
+            for rdcrop in range(10):
+                x, y = torch.randint(3000-512, (2,)).tolist()
```
[the remark this answers is not in the document]
Comment (Uniaut) [reply Dataloader/dataloader.py]: torch api로 random value를 줘야 seed 설정으로 결과를 통제할 수 있어서? 다른 좋은 방법 있으면 알려주ㅓ...

File: Dataloader/dataloader.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+            label = ele['properties']['type_id']
+            q = Quadangle(coords)
+
+            if img_id in self.features_map.keys():
+                self.features_map.get(img_id).add_box(q, label)
+            else:
+                # self.features_map[img_id] = Features(img_id)
+                pass
+
+        for img_id, f in self.features_map.items():
```
[the remark this answers is not in the document]
Comment (Uniaut) [reply Dataloader/dataloader.py]: f가 의미없어보인다는거지? ok

File: Dataloader/dataloader.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kozistr) [reply Dataloader/dataloader.py]: ㅇㅎ 따로 작성할 필요는 없을 듯

File: Dataloader/dataloader.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+            # print(img_id)
+            for rdcrop in range(10):
+                x, y = torch.randint(3000-512, (2,)).tolist()
+                # print((x, y, x + 512, y + 512))
+                img_info = ImageInfo(img_id, None, (x, y, x + 512, y + 512))
+                temp_f = copy.deepcopy(f)
```
[the remark this answers is not in the document]
Comment (Uniaut) [reply Dataloader/dataloader.py]: image에 있던 feature를 복제해서 따로 써야 했던건데 reference로 넘어가길래 막짜다가 이렇게됨
transform api 제대로 쓰면 이럴 일 없을듯

File: Dataloader/dataloader.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+                temp_f.crop_validate(img_info.crop)
+                self.items_map[img_info] = {'img_info': img_info, 'features': temp_f}
+
+        print(f'data load done. item length: {self.__len__()}')
+
+    def __len__(self):
+        return len(list(self.items_map.items()))
```
[the remark this answers is not in the document]
Comment (Uniaut) [reply Dataloader/dataloader.py]: 왜 이렇게 했을까?

File: Dataloader/dataloader.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kozistr) [reply Dataloader/dataloader.py]: ㅇㅎ random 값을 뽑는 법은 많이 있, 물론 seed 도 설정 가능하고

파이썬 내장 random 을 사용해도 되고 numpy randint 를 사용해도 되고 개인적으론 numpy randint 를 사용하는 걸 권장함

File: Dataloader/dataloader.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kozistr) [reply Dataloader/dataloader.py]: ㅇㅎ 그럼 일단 crop 하고 resize 부분은 최대한 torch API 로 리팩토링 ㄱㄱ

File: Dataloader/dataloader.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+start = time.time()
+
+ds = EOdata()
+
+print(time.time() - start)
+
+loader = dt.dataloader.DataLoader(dataset=ds,
+                                  batch_size=1,
+                                  shuffle=True,
+                                  collate_fn=temp_collate)
```
[the remark this answers is not in the document]
Comment (Uniaut) [reply Dataloader/dataloader.py]: 내가 output tensor 구조를 몰라서 이렇게 했음. 논문 읽어보고 다시 짤게

File: Dataloader/dataloader.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (kozistr) [reply Dataloader/dataloader.py]: ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ

File: Dataloader/dataloader.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Uniaut) [reply Dataloader/dataloader.py]: 베이스라인 모델도 rbox라는 이름 쓰던데 rbox로 바꿔야겠다

File: Dataloader/dataloader.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (kozistr) [reply Dataloader/dataloader.py]: 현재 dataset 이 `{'img': img, 'features': features}` 를 리턴하고 collate_fn 지나면

item[0]['img'], item[0]['features'] batch 에서 첫 번째 img, features 를 그냥 리턴하는데,

이건 그냥 dataset 에서 return img. features 하는 거랑 똑같아서 달아봤

File: Dataloader/dataloader.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Uniaut) [reply Dataloader/dataloader.py]: 인터넷에서 dict의 key로 custom class 쓰려면 이러래서....ㅎ;;

File: Dataloader/dataloader.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Uniaut) [reply Dataloader/dataloader.py]: torch에서 시드 관리해야 결과 통제되서 쓴건데 이렇게 쓰는게 아닌가?

File: Dataloader/dataloader.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Uniaut) [reply Dataloader/dataloader.py]: temp_f 에 내용 복제해서 쓰려고 했는데 계속 referencing으로 처리되어서 이렇게 만듬. transform api로 고치면 자연스럽게 없어질 코드일듯?

File: Dataloader/dataloader.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+    def __len__(self):
+        return len(list(self.items_map.items()))
+
+    def __getitem__(self, index):
+        items = list(self.items_map.values())[index]
+        img_info = items['img_info']
+        features = items['features']
+
+        if self.temp_img_id == img_info.img_id:
```
[the remark this answers is not in the document]
Comment (Uniaut) [reply Dataloader/dataloader.py]: 아 저거 같은 이미지 쓰는 item이 이웃하면 더 빠르게 돌아가도록 임시로 최적화 테스트해본거. 일단은 지우고 나중에 제대로 짜고 나서 최적화 할 때 물어볼게

File: Dataloader/dataloader.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Uniaut) [reply Dataloader/dataloader.py]: 이건 내가 output tensor 구성을 몰라서; 논문 읽어보고 제대로 짜겠음

File: Utils/Quadangle.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+import matplotlib
+from matplotlib import pyplot as plt
+from matplotlib.patches import Polygon
+from matplotlib.collections import PatchCollection
+
+
+class Quadangle:
+    def __init__(self, coords):
+        try:
+            if isinstance(coords, str):  # format: 'x1, y1, x2, ...' (unsorted)
```
[the remark this answers is not in the document]
Comment (Uniaut) [reply Utils/Quadangle.py]: todo) typing 에서 Union 찾아보기

File: Utils/Quadangle.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+                x1, y1 = coords[0], coords[1]
+                x2, y2 = coords[2], coords[3]
+                x3, y3 = coords[4], coords[5]
+                x4, y4 = coords[6], coords[7]
+
+                xc = (coords[0] + coords[2] + coords[4] + coords[6]) / 4
+                yc = (coords[1] + coords[3] + coords[5] + coords[7]) / 4
+
+                sorted_coords = [(x1, y1), (x2, y2), (x3, y3), (x4, y4)]
```
[the remark this answers is not in the document]
Comment (Uniaut) [reply Utils/Quadangle.py]: 이거 막짠 흔적이라 고쳐야됨 ㅎ;

File: Utils/Quadangle.py [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+            self.yc = (sorted_coords[0][1] + sorted_coords[1][1] + sorted_coords[2][1] + sorted_coords[3][1]) / 4
+
+            self.x1, self.y1 = sorted_coords[0][0], sorted_coords[0][1]
+            self.x2, self.y2 = sorted_coords[1][0], sorted_coords[1][1]
+            self.x3, self.y3 = sorted_coords[2][0], sorted_coords[2][1]
+            self.x4, self.y4 = sorted_coords[3][0], sorted_coords[3][1]
+
+            self.coords = sorted_coords
+
+        except IndexError:
```
[the remark this answers is not in the document]
Comment (Uniaut) [reply Utils/Quadangle.py]: 이거 원래 안일어나는데 code format 확인하려고 코드 작성 중간에 만듦. 지금은 필요없음

File: Utils/Quadangle.py [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+            print(f'This coordinate data is not appropriate: {coords}')
+
+    def __str__(self):
+        return f'QuadAngles<{self.coords}>'
+
+
+def print_q_if(q: Quadangle):
```
[the remark this answers is not in the document]
Comment (Uniaut) [reply Utils/Quadangle.py]: 이것도 테스트하려고 만든거라 필요없는 코드; 지워야 하는게 왜이렇게 많지

File: Dataloader/dataloader.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kozistr) [reply Dataloader/dataloader.py]: ㅇㅇㅇ 저렇게 사용도 가능하지

포인트는 구지 torch 로 랜덤을 뽑고 python list 로 변환해서 쓸 거면 그냥 파이썬 random 으로 seed 설정해 두고 randint 하는게 훨 나을 거 같아서 ㅇㅇ

np.random.seed(1337); np.random.randint() 이렇게 해도 reproducible 한 결과를 생성 가능함

File: Utils/Quadangle.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (kozistr) [reply Utils/Quadangle.py]: ㅇㅋ 수정 다 끝나면 review request 부분에 recycle 같은 버튼 눌러서 다시 불러주면 됨

File: Utils/Quadangle.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (kozistr) [reply Utils/Quadangle.py]: ㅇㅋ

File: Dataloader/dataloader.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kozistr) [reply Dataloader/dataloader.py]: 굳 굳이 rbox 말고도 너가 생각하는 적절한 이름 쓰면 될 듯

File: Dataloader/dataloader.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kozistr) [reply Dataloader/dataloader.py]: ㅇㅎ ㅇㅋ

File: Dataloader/dataloader.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kozistr) [reply Dataloader/dataloader.py]: ㅇㅋ

File: Dataloader/dataloader.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (kozistr) [reply Dataloader/dataloader.py]: ㅇㅋ

File: Dataloader/dataloader.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Uniaut) [reply Dataloader/dataloader.py]: +) 클래스 만든 목적이 음
데이터가 load나 inference 단계에선 coord 4개로 들어오는데 실제로 모델에선 x y w h theta 이렇게 집어넣어야 해서 train이랑 inference에서 자유롭게 변환하려고 만듦

[the remark this answers is not in the document]
Comment (kozistr) [reply Dataloader/dataloader.py]: ㅇㅎ 그런 목적이면 굳이 Class 로 만드는 것 보단 function level 로 만들어 봐도 좋을 듯

좌표 coord 4개 들어가면 theta 반환하는 함수, get_theta_from_coord 정도로 ㅇㅇ

File: Dataloader/dataloader.py [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+from PIL import Image, ImageDraw
+import torchvision.transforms
+import torch.random
+import copy
+
+from Utils.Quadangle import Quadangle
+
+
+PATH = 'C:/Devs/odship/dataset/'
+LABEL_PATH = PATH + 'labels.json'
```
[the remark this answers is not in the document]
Comment (Uniaut) [reply Dataloader/dataloader.py]: todo) http://pythonstudy.xyz/python/article/507-%ED%8C%8C%EC%9D%BC%EA%B3%BC-%EB%94%94%EB%A0%89%ED%86%A0%EB%A6%AC
